### PR TITLE
mfa: per-session MFA certs for SSH and Kubernetes

### PR DESCRIPTION
**Pull request by awly**
This is client-side support for requesting single-use certs with an MFA
check.

The client doesn't know whether they need MFA check when accessing a
resource, this is decided during an RBAC check on the server. So a
client will always try to get a single-use cert, and the server will
respond with NotNeeded if MFA is not required. This is an extra
round-trip for every session which causes ~20% slowdown in SSH logins:

```
$ hyperfine '/tmp/tsh-old ssh talos date' '/tmp/tsh-new ssh talos date'
Benchmark #1: /tmp/tsh-old ssh talos date
  Time (mean ± σ):      49.9 ms ±   1.0 ms    [User: 15.1 ms, System: 7.4 ms]
  Range (min … max):    48.4 ms …  54.1 ms    59 runs

Benchmark #2: /tmp/tsh-new ssh talos date
  Time (mean ± σ):      60.2 ms ±   1.6 ms    [User: 19.1 ms, System: 8.3 ms]
  Range (min … max):    59.0 ms …  69.7 ms    50 runs

  Warning: Statistical outliers were detected. Consider re-running this benchmark on a quiet PC without any interferences from other programs. It might help to use the '--warmup' or '--prepare' options.

Summary
  '/tmp/tsh-old ssh talos date' ran
    1.21 ± 0.04 times faster than '/tmp/tsh-new ssh talos date'
```

Another few other internal changes:

- client.LocalKeyAgent will now always have a non-nil LocalKeyStore.
  Previously, it would be nil (e.g. in a web UI handler or when using an
  identity file) which easily causes panics. I added a noLocalKeyStore
  type instead that returns errors from all methods.

- requesting a user cert with a TTL < 1min will now succeed and return a
  1min cert instead of failing

### CODE REVIEW

**Review by a-palchikov — APPROVED** [review]
I'm not certain about other ramifications of MFA check but if the 20% slowdown is a concern, the connect/port forward APIs could attempt to connect with the local agent's ssh key first (w/o the embedded node name) before requesting the certificate with MFA check if RBAC requires it.

**Comment by russjones** [comment]
@fspmarshall Can you review this?

File: lib/client/client.go [hunk 1]
```diff
@@ -1117,3 +1260,24 @@ func (proxy *ProxyClient) currentCluster() (*services.Site, error) {
...
+		return proxy.authMethod, nil
+	}
+	cluster, err := proxy.currentCluster()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	key, err := proxy.IssueUserCertsWithMFA(ctx, ReissueParams{
+		NodeName:       node,
+		RouteToCluster: cluster.Name,
+	})
```
Comment (fspmarshall) [inline lib/client/client.go:1277]: In order for this to work with the access request system, the request IDs need to be propagated (`ReissueParams.AccessRequests` field).  IDs of the currently active requests are stored on the SSH certificate.  #5441 introduces a helper method for extracting them, but you can do it manually with something like this:
```golang
var activeRequests services.RequestIDs
rawRequests, ok := sshCert.Extensions[teleport.CertExtensionTeleportActiveRequests]
if ok {
	if err := activeRequests.Unmarshal([]byte(rawRequests)); err != nil {
		return activeRequests, trace.Wrap(err)
	}
}
key, err := proxy.IssueUserCertsWithMFA(ctx, ReissueParams{
	AccessRequests: activeRequests.AccessRequests,
	// ...
})
```

File: lib/auth/grpcserver.go [hunk 2]
```diff
@@ -1603,26 +1618,115 @@ func (g *GRPCServer) GenerateUserSingleUseCerts(stream proto.AuthService_Generat
...
+			// Get the server address without port number.
+			addr, _, err := net.SplitHostPort(n.GetAddr())
+			if err != nil {
+				addr = n.GetAddr()
+			}
+			// Match NodeName to UUID, hostname or self-reported server address.
+			if n.GetName() == req.NodeName || n.GetHostname() == req.NodeName || addr == req.NodeName {
+				node = n
+				break
+			}
```
[the remark this answers is not in the document]
Comment (awly) [reply lib/auth/grpcserver.go:1656]: `n.GetName()` returns the UUID, so the first condition in this `if` statement matches nodes by UUID

File: lib/client/client.go (hunk 1 above)
Comment (awly) [reply lib/client/client.go:1293]: Thanks for the catch, added inside `IssueUserCertsWithMFA` to cover more callers.

File: lib/auth/grpcserver.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (fspmarshall) [reply lib/auth/grpcserver.go:1656]: Yeah, what I mean is that iteration shouldn't halt if a hostname/addr match is found because there might still be a node which matches UUID and that would need to take priority.

[the remark this answers is not in the document]
Comment (awly) [reply lib/auth/grpcserver.go]: If some node has a hostname matching other node's UUID (which seems very unlikely), then we have multiple matches and no way to resolve the conflict and for the user to pick one of them.

I think the more likely issue is when user passes a hostname and multiple nodes match.
Changed the logic to catch that scenario - if there are multiple matches return an error.

**Comment by awly** [comment]
PTAL @fspmarshall

**Review by fspmarshall — APPROVED** [review]

File: lib/auth/grpcserver.go [hunk 3]
```diff
@@ -1612,26 +1627,117 @@ func (g *GRPCServer) GenerateUserSingleUseCerts(stream proto.AuthService_Generat
...
 	}
-	maxExpiry := clock.Now().Add(teleport.UserSingleUseCertTTL)
+	// No error means that MFA is not required for this resource by
+	// AccessChecker.
+	if noMFAAccessErr == nil {
+		return errUserSingleUseUserCertNotNeeded
+	}
+	// Errors other than ErrSessionMFARequired mean something else is wrong,
+	// most likely access denied.
+	if noMFAAccessErr != services.ErrSessionMFARequired {
```
[the remark this answers is not in the document]
Comment (awly) [reply lib/auth/grpcserver.go]: good point.
I used `trace.Unwrap` originally and had trouble (because I think `trace.AccessDenied` unwraps into the underlying string error). But `errors.Is` works well, updated.

**Comment by awly** [comment]
@r0mant @russjones please review

File: lib/auth/grpcserver.go [hunk 4]
```diff
@@ -1612,26 +1627,116 @@ func (g *GRPCServer) GenerateUserSingleUseCerts(stream proto.AuthService_Generat
...
-func validateUserSingleUseCertRequest(req *proto.UserCertsRequest, clock clockwork.Clock) error {
+var errUserSingleUseUserCertNotNeeded = errors.New("single-use user certificate not needed to access this resource")
+
+// validateUserSingleUseCertRequest validates the request for a single-use user
+// cert. Returns nil if validation succeeds and MFA check is required, or
+// errUserSingleUseUserCertNotNeeded if validation succeeds and MFA
+// authentication is not required for this session. Any other error means
+// failed validation.
+func validateUserSingleUseCertRequest(ctx context.Context, actx *grpcContext, req *proto.UserCertsRequest) error {
+	if req.Username != actx.User.GetName() {
```
Comment (r0mant) [inline lib/auth/grpcserver.go:1638]: I wonder if you can use `actx.currentUserAction(req.Username)` here to not spread this authz logic across different servers (grpc vs server with roles)?

File: lib/auth/grpcserver.go [hunk 5]
```diff
@@ -1612,26 +1627,116 @@ func (g *GRPCServer) GenerateUserSingleUseCerts(stream proto.AuthService_Generat
...
+			if err != nil {
+				addr = n.GetAddr()
+			}
+			// Match NodeName to UUID, hostname or self-reported server address.
+			if n.GetName() == req.NodeName || n.GetHostname() == req.NodeName || addr == req.NodeName {
+				matches = append(matches, n)
+			}
+		}
+		if len(matches) == 0 {
+			return trace.NotFound("target resource not found")
```
Comment (r0mant) [inline lib/auth/grpcserver.go:1666]: Is the resource type/name not revealed in these messages on purpose?

File: lib/auth/grpcserver.go [hunk 6]
```diff
@@ -1612,26 +1627,116 @@ func (g *GRPCServer) GenerateUserSingleUseCerts(stream proto.AuthService_Generat
...
+var errUserSingleUseUserCertNotNeeded = errors.New("single-use user certificate not needed to access this resource")
+
+// validateUserSingleUseCertRequest validates the request for a single-use user
+// cert. Returns nil if validation succeeds and MFA check is required, or
+// errUserSingleUseUserCertNotNeeded if validation succeeds and MFA
+// authentication is not required for this session. Any other error means
+// failed validation.
+func validateUserSingleUseCertRequest(ctx context.Context, actx *grpcContext, req *proto.UserCertsRequest) error {
+	if req.Username != actx.User.GetName() {
+		return trace.BadParameter("cannot request a single-use certificate for a different user")
```
Comment (r0mant) [inline lib/auth/grpcserver.go:1639]: trace.AccessDenied?

File: lib/client/api.go [hunk 7]
```diff
@@ -1024,6 +1024,16 @@ func (tc *TeleportClient) ReissueUserCerts(ctx context.Context, params ReissuePa
 	return proxyClient.ReissueUserCerts(ctx, params)
 }
 
+func (tc *TeleportClient) IssueUserCertsWithMFA(ctx context.Context, params ReissueParams) (*Key, error) {
```
Comment (r0mant) [inline lib/client/api.go:1027]: Nit: Add godoc.

File: lib/client/mfa.go [hunk 8]
```diff
@@ -42,7 +42,7 @@ func PromptMFAChallenge(ctx context.Context, proxyAddr string, c *proto.MFAAuthe
 		return &proto.MFAAuthenticateResponse{}, nil
 	// TOTP only.
 	case c.TOTP != nil && len(c.U2F) == 0:
-		totpCode, err := prompt.Input(os.Stdout, os.Stdin, fmt.Sprintf("Enter an OTP code from a %sdevice", promptDevicePrefix))
+		totpCode, err := prompt.Input(os.Stderr, os.Stdin, fmt.Sprintf("Enter an OTP code from a %sdevice", promptDevicePrefix))
```
Comment (r0mant) [inline lib/client/mfa.go:45]: Just curious why this change to stderr?

File: lib/client/mfa.go [hunk 9]
```diff
@@ -126,6 +126,7 @@ func promptU2FChallenges(ctx context.Context, proxyAddr string, challenges []*pr
 		})
 	}
 
+	log.Debugf("prompting U2F devices with facet %q", facet)
```
Comment (r0mant) [inline lib/client/mfa.go:129]: Feels like a debug leftover, need this?

**Review by r0mant — APPROVED** [review]

File: lib/auth/grpcserver.go (hunk 5 above)
Comment (awly) [reply lib/auth/grpcserver.go]: I was just being lazy, updated

File: lib/client/mfa.go (hunk 8 above)
Comment (awly) [reply lib/client/mfa.go:45]: Because of `kubectl`'s exec plugins.
We configure `kubectl` to exec a `tsh` subcommand every time it needs to get credentials to connect.
`kubectl` expects to read the credentials from `stdout` of the execed process.
Anything written to `stderr` will be forwarded to `kubectl`'s own `stderr` to allow interactive commands like this.

File: lib/client/mfa.go (hunk 9 above)
Comment (awly) [reply lib/client/mfa.go:129]: This is left on purpose.
If a user mis-configures U2F facets (proxy addresses, basically) on the `auth_service`, it's really hard to debug when they use a different value in `tsh --proxy=...` and U2F auth will fail.

**Comment by awly** [comment]
Added `IsMFARequired` RPC on the auth server.
The client now uses this against the appropriate auth server (root or leaf) to check if MFA is required.
If it is, client sends the `UserSingleUseCerts` RPC to the **root** cluster, which should always be the one issuing credentials.

@r0mant @fspmarshall PTAL

File: api/client/proto/authservice.proto [hunk 10]
```diff
@@ -709,6 +709,25 @@ message UserSingleUseCertsResponse {
     }
 }
 
+// IsMFARequiredRequest is a request to check whether MFA is required to access
+// the Target.
+message IsMFARequiredRequest {
+    // Username of the caller.
+    string Username = 1;
```
Comment (fspmarshall) [inline api/client/proto/authservice.proto:716]: Clarify that this is OS user for target host, not teleport username.

**Review by fspmarshall — APPROVED** [review]

File: lib/auth/auth_with_roles.go [hunk 11]
```diff
@@ -66,7 +68,10 @@ func (a *ServerWithRoles) action(namespace string, resource string, action strin
 // even if they are not admins, e.g. update their own passwords,
 // or generate certificates, otherwise it will require admin privileges
 func (a *ServerWithRoles) currentUserAction(username string) error {
-	if a.hasLocalUserRole(a.context.Checker) && username == a.context.User.GetName() {
+	if hasLocalUserRole(a.context.Checker) && username == a.context.Identity.GetIdentity().Username {
+		return nil
+	}
+	if hasRemoteUserRole(a.context.Checker) && username == a.context.UnmappedIdentity.GetIdentity().Username {
```
Comment (r0mant) [inline lib/auth/auth_with_roles.go:74]: What's the implication of this change? Wouldn't this mean that, for example, a remote user has permissions to change the local user's password as long as they have the same username? I may be missing something.

File: lib/auth/auth_with_roles.go [hunk 12]
```diff
@@ -2601,6 +2611,132 @@ func (a *ServerWithRoles) GenerateUserSingleUseCerts(ctx context.Context) (proto
 	return nil, trace.NotImplemented("bug: GenerateUserSingleUseCerts must not be called on auth.ServerWithRoles")
 }
 
+func (a *ServerWithRoles) IsMFARequired(ctx context.Context, req *proto.IsMFARequiredRequest) (*proto.IsMFARequiredResponse, error) {
+	if err := a.currentUserAction(req.Username); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var noMFAAccessErr, notFoundErr error
+	switch t := req.Target.(type) {
```
Comment (r0mant) [inline lib/auth/auth_with_roles.go:2620]: More of a meta comment, but don't we want to keep the ACL layer slim without much business logic? Any downsides of moving this switch downstream, e.g. to an auth server method and invoke it here similar to how other methods here do? You can pass checker to it too.

File: lib/auth/auth_with_roles.go (hunk 11 above)
Comment (awly) [reply lib/auth/auth_with_roles.go]: Nice catch, didn't think about that.
It also turns out I was confusing "username" with "ssh login name", so I renamed that request field and it's no longer compared to the caller username.
Changes to `currentUserAction` reverted.

File: lib/auth/auth_with_roles.go (hunk 12 above)
Comment (awly) [reply lib/auth/auth_with_roles.go]: All of this code felt like pure authz logic, which is why I put it here.
But I think you're right, moved all of this code into `auth.Server`.

File: lib/client/client.go [hunk 13]
```diff
@@ -203,10 +246,156 @@ func (proxy *ProxyClient) ReissueUserCerts(ctx context.Context, params ReissuePa
...
+		return nil, trace.Wrap(err)
+	}
+	tlsCert, err := key.TeleportTLSCertificate()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	rootClusterName, err := tlsca.ClusterName(tlsCert.Issuer)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```
Comment (r0mant) [inline lib/client/client.go:270]: Nit: There's `RootClusterName()` method on the ProxyClient which you could probably just use instead of extracting from the TLS cert, although the outcome would probably be the same.

File: lib/client/client.go [hunk 14]
```diff
@@ -203,10 +246,156 @@ func (proxy *ProxyClient) ReissueUserCerts(ctx context.Context, params ReissuePa
...
+	if params.RouteToCluster == "" {
+		params.RouteToCluster = rootClusterName
+	}
+
+	// Note: always connect to root for getting new credentials.
+	clt, err := proxy.ConnectToCluster(ctx, params.RouteToCluster, true)
```
Comment (r0mant) [inline lib/client/client.go:276]: Shouldn't we use `proxy.ConnectToRootCluster()` here if we only ever want to connect to root? Or is the comment inaccurate?

File: lib/client/client.go [hunk 15]
```diff
@@ -203,10 +246,156 @@ func (proxy *ProxyClient) ReissueUserCerts(ctx context.Context, params ReissuePa
...
+			if err := activeRequests.Unmarshal([]byte(rawRequests)); err != nil {
+				return nil, trace.Wrap(err)
+			}
+		}
+		params.AccessRequests = activeRequests.AccessRequests
+	}
+
+	// Always connect to root for getting new credentials, but attempt to reuse
+	// the existing client if possible.
+	if params.RouteToCluster != rootClusterName {
```
Comment (r0mant) [inline lib/client/client.go:311]: If the comment above about connecting to root is accurate, then this is probably not needed?

File: lib/client/client.go [hunk 16]
```diff
@@ -203,10 +246,156 @@ func (proxy *ProxyClient) ReissueUserCerts(ctx context.Context, params ReissuePa
...
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	resp, err = stream.Recv()
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	certResp := resp.GetCert()
+	if cert == nil {
```
Comment (r0mant) [inline lib/client/client.go:379]: Should this be `if certResp == nil`?

**Review by r0mant — APPROVED** [review]
Couple more questions but otherwise lgtm.
